## Root-caused — it was mine, not the shard flakiness

The run finished and the log is retrievable now. Both reds are one root failure (`python-suite` is downstream of the shard):

```
autobot-slm-backend/slm/agent/health_collector_state_change_test.py:192
  → autobot-backend/services/notification_service.py:49
E   ModuleNotFoundError: No module named 'services.gateway'
```

**This PR caused it.** It added `from services.gateway.egress_governor import egress_governor` as a top-level import in `notification_service.py`. An slm-backend test loads that backend module **by path** and stubs its dependencies in `sys.modules` — and my new import was not in the stub list. Under the slm rootdir, `services` resolves to the *slm* package, so `services.gateway` is genuinely absent there.

I had flagged #14111 as the likely explanation. It was not — checking rather than assuming was the difference.

## Fix

Added the two missing stubs alongside the existing ones, with a comment explaining why every new backend top-level import has to be mirrored there.

**Verified against the reproduction, not the predicate.** The failure never appears from the repo root, which is why it looked CI-only. Running with the slm rootdir reproduces it exactly:

```
$ cd autobot-slm-backend && pytest slm/agent/health_collector_state_change_test.py
E   ModuleNotFoundError: No module named 'services.gateway'; 'services' is not a package
    2 failed  ← same two tests as CI

$ # with the stubs restored, same invocation
    14 passed
```

Pushed as `058348d31`, rebased onto the current branch tip.

--- a/autobot-backend/api/integration_communication.py
+++ b/autobot-backend/api/integration_communication.py
@@ -282,6 +282,16 @@ async def send_webhook_message(
     )
     integration = TeamsIntegration(config)
 
+    # #14270: this endpoint is a *sibling* of `send_message` above, not a branch
+    # of it. The comment there covers the Teams fallback inside that function and
+    # says nothing about this one, which reaches the same Teams webhook by its
+    # own route — so governing that function left this bypass wide open, reading
+    # as covered because the governor appears elsewhere in the module.
+    verdict = await egress_governor.evaluate(platform="teams", channel_id="", message_id="")
+    if not verdict.allowed:
+        logger.warning("teams webhook send blocked by egress governance (%s): %s", verdict.rule, verdict.reason)
+        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Outbound send denied by egress policy")
+
     try:
         params = {"text": webhook.text}
         if webhook.title:

--- a/autobot-backend/services/gateway/live_egress_seams_test.py
+++ b/autobot-backend/services/gateway/live_egress_seams_test.py
@@ -285,3 +285,43 @@ class TestTheSiblingSendersReviewFound:
 
         spy.assert_awaited(), "the email seam sent without an audit record"
         assert spy.await_args.kwargs["platform"] == "email"
+
+
+class TestTheNotificationSeamsHonourADenial:
+    """Flagged by security review as fail-open gates.
+
+    Both notification seams awaited the governor and discarded the verdict.
+    That is *currently* equivalent to checking it — `_decide` returns allowed
+    unconditionally when `require_approval=False` — so it was not a reachable
+    bypass. It was a dependency on the governor's present shape that nothing in
+    the calling file recorded, which is the same trap as reusing a flag that
+    happens to answer a different question.
+
+    These tests force a denial through, so the seams cannot go back to ignoring
+    one if `_decide` ever grows a non-approval rule.
+    """
+
+    @pytest.mark.asyncio
+    async def test_a_denied_verdict_stops_the_email(self):
+        from services.notification_service import NotificationService
+
+        svc = NotificationService()
+        with patch("services.notification_service.egress_governor.evaluate", new=AsyncMock(side_effect=_deny)):
+            with patch("smtplib.SMTP") as smtp, patch("smtplib.SMTP_SSL") as smtp_ssl:
+                await svc._send_email("someone@example.invalid", "subj", "body")
+
+        smtp.assert_not_called()
+        smtp_ssl.assert_not_called()
+
+    @pytest.mark.asyncio
+    async def test_a_denied_verdict_stops_the_webhook_post(self):
+        from services.notification_service import NotificationService
+
+        svc = NotificationService()
+        http = MagicMock()
+        with patch("services.notification_service.egress_governor.evaluate", new=AsyncMock(side_effect=_deny)):
+            with patch("autobot_shared.http_client.get_http_client", return_value=http) as client:
+                await svc._send_webhook("https://hooks.example.invalid/x", {"t": 1})
+
+        client.assert_not_called()
+        http.tracked_request.assert_not_called()

--- a/autobot-backend/services/gateway/live_egress_seams_test.py
+++ b/autobot-backend/services/gateway/live_egress_seams_test.py
@@ -13,7 +13,7 @@ Every test here drives the **producer**, not the governor. A test that asserts
 patch the governor to deny and assert nothing reaches the transport.
 """
 
-from unittest.mock import AsyncMock, patch
+from unittest.mock import AsyncMock, MagicMock, patch
 
 import pytest
 
@@ -211,3 +211,77 @@ class TestNotificationAlertsAreAuditedButNeverBlocked:
 
         assert spy.await_args.kwargs["channel_id"] == "hooks.example.invalid"
         assert "secret-token-path" not in spy.await_args.kwargs["channel_id"]
+
+
+class TestTheSiblingSendersReviewFound:
+    """The gaps that shipped in the first version of #14270.
+
+    Every seam below reaches a real recipient and was ungoverned while a sibling
+    in the same file was carefully gated — which is the shape that makes a
+    bypass invisible: the governor's name appears in the module, so a reader
+    checking for coverage finds it and stops.
+
+    `send_photo` and `send_document` matter most: both are reachable from the
+    agent-response dispatcher this control exists to protect, so an agent
+    answering with an image bypassed the gate entirely.
+    """
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize(
+        "method,kwargs",
+        [
+            ("send_photo", {"chat_id": "c-1", "photo": "file-id"}),
+            ("send_document", {"chat_id": "c-1", "document": "file-id"}),
+        ],
+    )
+    async def test_a_denied_verdict_stops_the_telegram_upload(self, method, kwargs):
+        from services.telegram_bot_service import TelegramBotService
+
+        svc = TelegramBotService(bot_token="t")
+        http = MagicMock()
+        with patch("services.telegram_bot_service.egress_governor.evaluate", new=AsyncMock(side_effect=_deny)):
+            with patch("services.telegram_bot_service.get_http_client", return_value=http) as client:
+                result = await getattr(svc, method)(**kwargs)
+
+        assert result["error"] == "egress_denied"
+        client.assert_not_called()
+        http.tracked_request.assert_not_called()
+
+    @pytest.mark.asyncio
+    async def test_a_denied_verdict_stops_the_teams_webhook_endpoint(self):
+        """`send_webhook_message` is a sibling route, not a branch of
+        `send_message` — governing that function left this one open."""
+        from fastapi import HTTPException
+
+        from api.integration_communication import send_webhook_message
+
+        integration = MagicMock()
+        integration.execute_action = AsyncMock()
+        with patch("api.integration_communication.egress_governor.evaluate", new=AsyncMock(side_effect=_deny)):
+            with patch("api.integration_communication.TeamsIntegration", return_value=integration):
+                with pytest.raises(HTTPException) as raised:
+                    await send_webhook_message(
+                        provider="teams",
+                        webhook=MagicMock(text="hello", title=None, webhook_url="https://example.invalid/hook"),
+                    )
+
+        assert raised.value.status_code == 403
+        integration.execute_action.assert_not_awaited()
+
+    @pytest.mark.asyncio
+    async def test_the_email_seam_is_audited(self):
+        """SMTP reaches a real external recipient. It was the one notification
+        channel left out while its siblings were wired."""
+        from services.notification_service import NotificationService
+
+        svc = NotificationService()
+        spy = AsyncMock(side_effect=_allow)
+        with patch("services.notification_service.egress_governor.evaluate", new=spy):
+            with patch("smtplib.SMTP"), patch("smtplib.SMTP_SSL"):
+                try:
+                    await svc._send_email("someone@example.invalid", "subj", "body")
+                except Exception:
+                    pass
+
+        spy.assert_awaited(), "the email seam sent without an audit record"
+        assert spy.await_args.kwargs["platform"] == "email"

--- a/autobot-backend/services/notification_service.py
+++ b/autobot-backend/services/notification_service.py
@@ -444,12 +444,22 @@ class NotificationService:
         # the control look complete while one door stayed open.
         # require_approval=False for the same reason as the other alert channels:
         # an APPROVAL_NEEDED alert must not be gated by the approval system.
-        await egress_governor.evaluate(
+        #
+        # The verdict is still checked. With require_approval=False the governor
+        # currently always allows, so this reads as dead — it is not. It stops
+        # the seam depending on `_decide`'s present shape: a policy that denies
+        # for a reason other than approval (a blocklist, a kill switch) would
+        # otherwise be silently ignored here, and nothing in this file would say
+        # so. Two reviews flagged the discarded verdict as a fail-open gate.
+        verdict = await egress_governor.evaluate(
             platform="email",
             channel_id="",
             message_id="",
             require_approval=False,
         )
+        if not verdict.allowed:
+            logger.warning("email notification blocked by egress governance (%s): %s", verdict.rule, verdict.reason)
+            return
 
         try:
             if use_tls:
@@ -486,12 +496,15 @@ class NotificationService:
         # explicitly rather than relying on the env default, so arming the policy
         # cannot silence outage alerts or deadlock APPROVAL_NEEDED. The record
         # still lands, so an operator can see what left the machine.
-        await egress_governor.evaluate(
+        verdict = await egress_governor.evaluate(
             platform="webhook",
             channel_id=urlparse(url).hostname or "unknown",
             message_id="",
             require_approval=False,
         )
+        if not verdict.allowed:
+            logger.warning("webhook notification blocked by egress governance (%s): %s", verdict.rule, verdict.reason)
+            return
 
         headers = {
             "Content-Type": "application/json",

--- a/autobot-backend/services/notification_service.py
+++ b/autobot-backend/services/notification_service.py
@@ -438,6 +438,19 @@ class NotificationService:
         msg["To"] = to
         msg.attach(MIMEText(body, "plain", "utf-8"))
 
+        # #14270: email was the one notification channel left ungated while its
+        # siblings were wired. SMTP delivers to a real external recipient — the
+        # same class of action as the webhook path above — so leaving it out made
+        # the control look complete while one door stayed open.
+        # require_approval=False for the same reason as the other alert channels:
+        # an APPROVAL_NEEDED alert must not be gated by the approval system.
+        await egress_governor.evaluate(
+            platform="email",
+            channel_id="",
+            message_id="",
+            require_approval=False,
+        )
+
         try:
             if use_tls:
                 from autobot_shared.tls import get_internal_tls_context

--- a/autobot-backend/services/telegram_bot_service.py
+++ b/autobot-backend/services/telegram_bot_service.py
@@ -79,6 +79,34 @@ class TelegramBotService:
 
         return cls(bot_token=bot_token)
 
+    async def _egress_denied(
+        self,
+        chat_id: str,
+        reply_to_message_id: Optional[int] = None,
+        require_approval: Optional[bool] = None,
+    ) -> Optional[Dict[str, Any]]:
+        """The egress gate every live Telegram send passes through (#14270).
+
+        One helper rather than the check inlined per method: this class has three
+        senders that each reach the Bot API, and the first version of #14270
+        gated only ``send_message``. Review found ``send_photo`` and
+        ``send_document`` still posting ungoverned — both reachable from the
+        agent-response dispatcher this control exists to protect. A copied guard
+        is a guard the next sender silently omits.
+
+        Returns the denial response to hand straight back, or ``None`` to proceed.
+        """
+        verdict = await egress_governor.evaluate(
+            platform="telegram",
+            channel_id=str(chat_id),
+            message_id=str(reply_to_message_id or ""),
+            require_approval=require_approval,
+        )
+        if verdict.allowed:
+            return None
+        logger.warning("Telegram send blocked by egress governance (%s): %s", verdict.rule, verdict.reason)
+        return {"ok": False, "error": "egress_denied", "reason": verdict.reason}
+
     async def send_message(
         self,
         chat_id: str,
@@ -122,15 +150,9 @@ class TelegramBotService:
         # a failure — APPROVAL_NEEDED above all — must never be gated by the
         # approval system it exists to serve, or arming the policy deadlocks it.
         # Callers opt into that explicitly; the default (None) follows the env flag.
-        verdict = await egress_governor.evaluate(
-            platform="telegram",
-            channel_id=str(chat_id),
-            message_id=str(reply_to_message_id or ""),
-            require_approval=require_approval,
-        )
-        if not verdict.allowed:
-            logger.warning("Telegram send blocked by egress governance (%s): %s", verdict.rule, verdict.reason)
-            return {"ok": False, "error": "egress_denied", "reason": verdict.reason}
+        denied = await self._egress_denied(chat_id, reply_to_message_id, require_approval)
+        if denied:
+            return denied
 
         url = f"{self.base_url}/sendMessage"
         async with get_http_client().tracked_request("POST", url, json=payload) as response:
@@ -302,6 +324,10 @@ class TelegramBotService:
         if message_thread_id:
             payload["message_thread_id"] = message_thread_id
 
+        denied = await self._egress_denied(chat_id, reply_to_message_id)
+        if denied:
+            return denied
+
         url = f"{self.base_url}/sendPhoto"
         async with get_http_client().tracked_request("POST", url, json=payload) as response:
             if response.status != 200:
@@ -355,6 +381,10 @@ class TelegramBotService:
 
         if message_thread_id:
             payload["message_thread_id"] = message_thread_id
+
+        denied = await self._egress_denied(chat_id, reply_to_message_id)
+        if denied:
+            return denied
 
         url = f"{self.base_url}/sendDocument"
         async with get_http_client().tracked_request("POST", url, json=payload) as response:

--- a/changelog/unreleased/14270-live-egress-seams.md
+++ b/changelog/unreleased/14270-live-egress-seams.md
@@ -5,3 +5,7 @@ issue: 14270
 pr: 0
 ---
 Outbound sends on the live channels (Telegram, WhatsApp, Slack/Discord/Teams) now pass the egress governance stage, so every message to a real person is recorded and can be gated; operational alerts are audited but never blocked.
+- Cover the sibling senders review found ungoverned: Telegram `send_photo` and
+  `send_document` (both reachable from the agent-response dispatcher), the Teams
+  webhook endpoint, and the SMTP notification channel. The Telegram gate is now
+  one helper rather than a check copied per method.

